scripts/release/main.go

```diff
@@ -0,0 +1,223 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"slices"
+	"strings"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-github/v61/github"
+	"golang.org/x/mod/semver"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+	"cdr.dev/slog/sloggers/sloghuman"
+	"github.com/coder/coder/v2/cli/cliui"
+	"github.com/coder/serpent"
+)
+
+const (
+	owner = "coder"
+	repo  = "coder"
+)
+
+func main() {
+	logger := slog.Make(sloghuman.Sink(os.Stderr)).Leveled(slog.LevelDebug)
+
+	var ghToken string
+	var dryRun bool
+
+	cmd := serpent.Command{
+		Use:   "release <subcommand>",
+		Short: "Prepare, create and publish releases.",
+		Options: serpent.OptionSet{
+			{
+				Flag:        "gh-token",
+				Description: "GitHub personal access token.",
+				Env:         "GH_TOKEN",
+				Value:       serpent.StringOf(&ghToken),
+			},
+			{
+				Flag:          "dry-run",
+				FlagShorthand: "n",
+				Description:   "Do not make any changes, only print what would be done.",
+				Value:         serpent.BoolOf(&dryRun),
+			},
+		},
+		Children: []*serpent.Command{
+			{
+				Use:   "promote <version>",
+				Short: "Promote version to stable.",
+				Handler: func(inv *serpent.Invocation) error {
+					ctx := inv.Context()
+					if len(inv.Args) == 0 {
+						return xerrors.New("version argument missing")
+					}
+					if !dryRun && ghToken == "" {
+						return xerrors.New("GitHub personal access token is required, use --gh-token or GH_TOKEN")
+					}
+
+					err := promoteVersionToStable(ctx, inv, logger, ghToken, dryRun, inv.Args[0])
+					if err != nil {
+						return err
+					}
+
+					return nil
+				},
+			},
+		},
+	}
+
+	err := cmd.Invoke().WithOS().Run()
+	if err != nil {
+		if errors.Is(err, cliui.Canceled) {
+			os.Exit(1)
+		}
+		logger.Error(context.Background(), "release command failed", "err", err)
+		os.Exit(1)
+	}
+}
+
+//nolint:revive // Allow dryRun control flag.
+func promoteVersionToStable(ctx context.Context, inv *serpent.Invocation, logger slog.Logger, ghToken string, dryRun bool, version string) error {
+	client := github.NewClient(nil)
+	if ghToken != "" {
+		client = client.WithAuthToken(ghToken)
+	}
+
+	logger = logger.With(slog.F("dry_run", dryRun), slog.F("version", version))
+
+	logger.Info(ctx, "checking current stable release")
+
+	// Check if the version is already the latest stable release.
+	currentStable, _, err := client.Repositories.GetLatestRelease(ctx, "coder", "coder")
+	if err != nil {
+		return err
+	}
+
+	logger = logger.With(slog.F("stable_version", currentStable.GetTagName()))
+	logger.Info(ctx, "found current stable release")
+
+	if currentStable.GetTagName() == version {
+		return xerrors.Errorf("version %q is already the latest stable release", version)
+	}
+
+	// Ensure the version is a valid release.
+	perPage := 20
+	latestReleases, _, err := client.Repositories.ListReleases(ctx, owner, repo, &github.ListOptions{
+		Page:    0,
+		PerPage: perPage,
+	})
+	if err != nil {
+		return err
```

Comment:
I'd suggest wrapping this error or adding a log line before it showing what the current action is